### PR TITLE
Components: Remove "experimental" designation

### PR DESCRIPTION
**Pull request by fullofcaffeine**
Solves: https://github.com/WordPress/gutenberg/issues/59418.

## What?

Remove `__experimental` prefix from all "experimental" components, effectively promoting them to regular stable components. See the related issue for more context.

Note: Some more complex conversions could be done in a separate PR, like the one for `CustomSelectControl,` which also requires migrating to a new adapter component for current consumers.

## Why?

The strategy of prefixing exports with `__experimental` has become deprecated after the introduction of private APIs. Read more in the related issue. 

## How?

For each `__experimental` component:

1) Exports it from components without the `__experimental` prefix;
2) Keep the old `__experimental` export for backwards compatibility;
4) Change all imports of the old `__experimental` in GB and components to the one without the prefix (including in storybook stories). Also update the docs to refer to the new unprefixed component;
5) Add the component id to the `PREVIOUSLY_EXPERIMENTAL_COMPONENTS` const array in `manager-head.html` so that old `experimental` story paths are redirected to the new one.

**TODO:**
- [ ] Add changelog entry(ies) about the removal the `__experimental` prefix. Perhaps `Components: Remove "experimental" designation for <component>` for each component would be the best changelog entry, semantically?

## Testing Instructions

* All checks should pass;
* Unit tests and E2Es should pass;
* Find where the consumers of those components are and smoke test them.

### CODE REVIEW

File: packages/components/src/index.ts [hunk 1]
```diff
@@ -12,7 +12,10 @@ export {
 } from '@wordpress/primitives';
 
 // Components.
-export { default as __experimentalAlignmentMatrixControl } from './alignment-matrix-control';
+export {
+	default as __experimentalAlignmentMatrixControl,
```
Comment (tyxla) [inline packages/components/src/index.ts:16]: While I appreciate that we're keeping the `__experimental` exports for backward compatibility reasons, I wonder if we should deprecate `__experimental` imports now that the components are stable. That will nudge 3rd party developers to migrate to stable imports, otherwise they might be stuck with the `__experimental` ones forever

File: packages/block-library/src/navigation/edit/navigation-menu-delete-control.js [hunk 2]
```diff
@@ -1,10 +1,7 @@
 /**
  * WordPress dependencies
  */
-import {
-	Button,
-	__experimentalConfirmDialog as ConfirmDialog,
-} from '@wordpress/components';
+import { Button, ConfirmDialog } from '@wordpress/components';
```
Comment (tyxla) [inline packages/block-library/src/navigation/edit/navigation-menu-delete-control.js:4]: I think updating all of Gutenberg to use the stable component imports can be its own PR. There are quite a few of them that this PR isn't covering just yet, and splitting to smaller PRs could help with reviewing.

**Review by tyxla — COMMENTED** [review]
Thanks for working on it @fullofcaffeine 👍 

I know you're still working on it, but sending some early high-level feedback, in case it's helpful.

File: packages/components/src/index.ts (hunk 1 above)
Comment (mirka) [reply packages/components/src/index.ts:16]: There are so many "shades" of deprecation — hard, soft, TS-only, TSDoc only... 😂 For these I think a TSDoc <code>@deprecated Import `AlignmentMatrixControl` instead</code> would be appropriate. Is that what you had in mind?

File: packages/block-library/src/navigation/edit/navigation-menu-delete-control.js (hunk 2 above)
Comment (mirka) [reply packages/block-library/src/navigation/edit/navigation-menu-delete-control.js:4]: Doing it in a separate PR would also help to check that we didn't remove any `__experimental` exports by mistake.

File: packages/components/src/index.ts [hunk 3]
```diff
@@ -36,6 +36,10 @@ export {
 	hasSplitBorders as __experimentalHasSplitBorders,
 	isDefinedBorder as __experimentalIsDefinedBorder,
 	isEmptyBorder as __experimentalIsEmptyBorder,
+	BorderBoxControl,
+	hasSplitBorders,
+	isDefinedBorder,
+	isEmptyBorder,
```
Comment (mirka) [inline packages/components/src/index.ts:42]: Ok, these three exports are sounding major alarm bells for me 😱 Not sure what's going on there. Let's keep these experimental for now so we can investigate and figure out a better way.

File: packages/components/src/index.native.js
Comment (mirka) [inline packages/components/src/index.native.js:1]: Let's not touch the `*.native.js` stuff for now, they are completely separate.

**Review by mirka — COMMENTED** [review]
I reviewed the commits for the first two components, AlignmentMatrixControl and BorderBoxControl.

The checklist of what needs to be done for each component looks good, pending the decision about if/how to flag as deprecated. One more thing we need to do is to remove these callouts in the readmes:

https://github.com/WordPress/gutenberg/blob/752a07f550aaf3c9a59c9fdf65679bd7688bbc48/packages/components/src/alignment-matrix-control/README.md?plain=1#L3-L5

---

On another note, I'm sure this de-experimentalizing process will reveal some smells, where we'll need to make case-by-case decisions. One example I already saw is the assortment of utility functions being exported in relation to BorderBoxControl. Feel free to flag if you notice anything fishy like that, for example components that aren't being used anywhere in Gutenberg.

So given that this process does require relatively careful review, I'm thinking we can chunk PRs into several components each, so we can better detect abnormalities. Does that sound about right?

File: packages/components/src/index.ts (hunk 1 above)
Comment (tyxla) [reply packages/components/src/index.ts:16]: Yup, it's odd that deprecation can be so complex 😅 

Anyway, imagine I'm a plugin dev that uses some experimental components in my plugin. How will a TSDoc `@deprecated` help me find out that I'm using a deprecated import and I have to replace it with the canonical stable import? Will I even be notified about it by my editor, or does that rely on me being very attentive to reading the type hints provided by my editor?

File: packages/block-library/src/navigation/edit/navigation-menu-delete-control.js (hunk 2 above)
Comment (fullofcaffeine) [reply packages/block-library/src/navigation/edit/navigation-menu-delete-control.js:4]: Sounds good! I'll rework that 👍🏻

File: packages/components/src/index.ts (hunk 1 above)
Comment (fullofcaffeine) [reply packages/components/src/index.ts:16]: > Will I even be notified about it by my editor, or does that rely on me being very attentive to reading the type hints provided by my editor?

I managed to make the `@deprecate` annotation work when it's added just before the actual export symbol, inside the `export` block:

```typescript
export {
	/**
	 * @deprecated Since version X.X. Will be removed in version X.Y. Use `ConfirmDialog` instead.
	 */
	ConfirmDialog as __experimentalConfirmDialog,
	ConfirmDialog,
} from './confirm-dialog';
```

I can confirm it doesn't affect the second export there:

Screenshot 2024-04-18 at 12 51 30 p m

And of the `@deprecated` declaration, it will render like this in `vscode` (notice the strikethrough in the import and the warning in the IntelliSense popup):

| | |
|-|-|
| Screenshot 2024-04-18 at 12 48 28 p m | Screenshot 2024-04-18 at 12 49 04 p m |

So I assume all editors/IDEs that support tsdoc will warn users about it.

If that's a good way to prevent developers from continuing to use it, that is another story. It might be good to do the deprecation gradually? First soft-deprecate using a tsdoc, announce in the changelog and perhaps a Make post, then in another (major?) version, completely remove the `__experimental` exports.

Comment (tyxla) [reply packages/components/src/index.ts:16]: Nice 👍 

And yeah, gradual deprecation sounds good to me. We should likely document the plan for that in the related issue, including some releases where we aim to further/hard deprecate.

**Comment by fullofcaffeine** [comment]
Closed in favor of https://github.com/WordPress/gutenberg/pull/60884, where I focused on removing the `experimental` category for components only. It was easier to start in a new PR than changing this one. I'll also keep this as a reference. 

I'll follow up later with a PR to replace the imports and another to work on replacing the exported implementation of `CUstomSelectControl` to `V2` (which I started looking into in this WIP commit: https://github.com/WordPress/gutenberg/pull/60837/commits/752a07f550aaf3c9a59c9fdf65679bd7688bbc48).

File: packages/components/src/index.ts (hunk 1 above)
Comment (mirka) [reply packages/components/src/index.ts:16]: > Anyway, imagine I'm a plugin dev that uses some experimental components in my plugin. How will a TSDoc `@deprecated` help me find out that I'm using a deprecated import and I have to replace it with the canonical stable import? Will I even be notified about it by my editor, or does that rely on me being very attentive to reading the type hints provided by my editor?

My current stance on this is, 

1. Do we plan to hard deprecate it?
2. Does switching to the new thing provide the consumer with actual benefits over the old thing?

If the answer is "no" to both questions, we shouldn't log a console warning and we shouldn't block the TS from compiling. We shouldn't demand a dev's attention for what is essentially a housekeeping detail. If they care about housekeeping details, they'll do it.

And to be clear, personally I don't think the `__experimental` exports are worth hard deprecating. It's virtually free to maintain this BC, compared to the cost on the ecosystem if we yank them.

Comment (tyxla) [reply packages/components/src/index.ts:16]: Oh, the complex deprecation discussion again 😅 

I think we have far too many cases that are cheap/free to maintain, but because we have too many, it adds so much noise. I wish we had a clear deprecation policy that actually allows us to get rid of deprecated features at some point. 

If I come back to planet Earth though and draw a line on the tradeoffs, you're right and it should be just fine to continue soft-deprecating as suggested.

File: packages/components/src/index.ts (hunk 3 above)
Comment (DaniGuardiola) [reply packages/components/src/index.ts:42]: See: https://github.com/WordPress/gutenberg/issues/61135
